Add associations, migrations, validations and mailer

ritika1206 opened the pull request:
User can purchase netflix subscription

Upto 4 users can share a netflix subscription

Each subscription is valid for some time

All the important data should be accessible through associations.

User should not be able to have more than one active(not expired) subscription associated with it

Send a reminder email to all the users of the subscription a month, a week and a day before the subscription expires with the appropriate email message.

Commit:
Add associations, migrations, validations and mailer

apekshit-tewathia commented on app/models/subscription.rb:
```diff
@@ -0,0 +1,16 @@
+class Subscription < ApplicationRecord
+  has_many :user_ids, class_name: 'UserSubscription', dependent: :destroy, counter_cache: :users_count, foreign_key: :subsciption_id, inverse_of: :subscription
```
rename

apekshit-tewathia commented on app/models/subscription.rb:
```diff
@@ -0,0 +1,16 @@
+  has_many :users, through: :user_ids, source: :user
+
+  validates :name, :price, :validity, presence: true
+  validates :users_count, numericality: { less_than_or_equal_to: 4 }
```
4 in a constant

apekshit-tewathia commented on app/models/user.rb:
```diff
@@ -0,0 +1,16 @@
+class User < ApplicationRecord
+  has_many :subsciption_ids, class_name: 'UserSubscription', dependent: :destroy, counter_cache: :subscriptions_count, foreign_key: :user_id, inverse_of: :user 
```
let's simply name this as user_subscriptions so we don't have to pass in so many options

apekshit-tewathia commented on app/models/user.rb:
```diff
@@ -0,0 +1,16 @@
+
+  validates :name, :email, presence: true
+  validates :name, format: { with: Netflix::REGEX_NAME }, allow_blank: true
+  validates :active_subscriptions_count, numericality: { is: 1 }
```
constant

apekshit-tewathia commented on app/models/user.rb:
```diff
@@ -0,0 +1,16 @@
+  private
+
+    def active_subscriptions_count
+      subsciptions.where(status: 'active').count
```
create an association named active_subscriptions and call count on that

Commit:
Resolved PR comments-
- Add validity days to months and weeks conversion helper
- Create expiration_date method
- Create has_one :active_subscription association

Commit:
minor change

Commit:
Constant extraction

Commit:
Resolve PR comments changes
